src/script/interpreter.cpp

```diff
@@ -1110,35 +1110,46 @@ class CTransactionSignatureSerializer {
 
 } // anon namespace
 
-uint256 SignatureHash(const CScript& scriptCode, const CTransaction& txTo, unsigned int nIn, int nHashType, const CAmount& amount, int sigversion)
+uint256 SignatureHash(const CScript& scriptCode, const CTransaction& txTo, unsigned int nIn, int nHashType, const CAmount& amount, int sigversion, CachedHashes* cache)
 {
+    CachedHashes empty;
+    if(cache == NULL)
+    {
+        cache = &empty;
+    }
+    
     if (sigversion == 1) {
+        if(cache->IsEmpty())
```

Comment:
This is actually quite complex.
My first attempt was https://github.com/NicolasDorier/bitcoin/blob/8319c4a392087f43fcb7806eeba633b907bdfb91/src/script/interpreter.cpp#L1077

The code is more complex, but worse, there was a bug.
Imagine first input was AnyoneCanPay. Then the CachedHashes would not have `hashSequence` set. However, the CachedHashes instance would be in the cache.

If the second input is not AnyoneCanPay, the code would take this incomplete CachedHashes out of the map and calculate `hashOutputs` as it should.
However, since the caller took the CachedHashes out of the cache already, it will not put the new value into it. (because I want to minimize the lock access)

So for the following inputs, each time, the hashOutput would be recalculated from scratch.

Of course, this would be fixable, but would make the code more difficult to review for not a lot of upside.